File: typescript/tests/e2e/smoke.ts
import { emit, entrypoint, step, workflow } from "../../src/index.js";

type SmokeInput = {
  iterations?: number;
};

const smoke = workflow<SmokeInput, { ok: boolean; total: number; iterations: number }>(
  "e2e-smoke-ts",
  async (input = {}) => {
    const iterations = input.iterations ?? 3;
    const results: Array<{ val: number }> = [];

    for (let i = 0; i < iterations; i++) {
      const result = await step(
        `step-${i}`,
        { index: i },
        async () => ({ val: i * 2 }),
      );
      results.push(result);
    }

    const total = results.reduce((sum, r) => sum + r.val, 0);

    await emit("total", total);
    await emit("iterations", iterations);

    return { ok: true, total, iterations };
  },
);

entrypoint(smoke);

File: python/tests/e2e/smoke.py
from collections.abc import Awaitable, Callable
from typing import cast

from quantiles import emit, entrypoint, step, workflow
from quantiles.types import JsonValue
from quantiles.workflow_context import WorkflowContext


async def _smoke_handler(_input: JsonValue, ctx: WorkflowContext) -> JsonValue:
  iterations = 3
  if isinstance(_input, dict):
    raw = _input.get("iterations")
    if isinstance(raw, int):
      iterations = raw

  results: list[JsonValue] = []
  for i in range(iterations):
    _input_value: JsonValue = {"index": i}

    def _make_execute(idx: int) -> Callable[[], Awaitable[JsonValue]]:
      async def _execute() -> JsonValue:
        return await _make_step_result(idx)

      return _execute

    result = await step(
      ctx,
      step_key=f"step-{i}",
      input_value=_input_value,
      execute=_make_execute(i),
    )
    results.append(result)

  total = 0
  for r in results:
    val = cast(dict[str, JsonValue], r).get("val", 0)
    if isinstance(val, int):
      total += val

  await emit(ctx, "total", float(total))
  await emit(ctx, "iterations", float(iterations))

  return cast(JsonValue, {"ok": True, "total": total, "iterations": iterations})


async def _make_step_result(idx: int) -> JsonValue:
  return {"val": idx * 2}


smoke = workflow("e2e-smoke-py", _smoke_handler)
entrypoint(smoke)
